force-app/main/default/classes/DMLOperations.cls

```diff
@@ -143,12 +174,55 @@ public with sharing class DMLOperations {
      * @param accountName The name of the Account.
      * @param oppNames The list of Opportunity names.
      */
-    public static void upsertOpportunities(String accountName, List<String> oppNames) {
-        // Loop through each Opportunity name
-            // Create an Opportunity with the given name
-            // Add the Opportunity to the list
-    }
     
+     public static void upsertOpportunities(String accountName, List<String> oppNames) {
+         // Query for the account
+        Account account;
+        List<Account> accounts = [SELECT Id FROM Account WHERE Name = :accountName LIMIT 1];
+        if (accounts.isEmpty()) {
+            account = new Account(Name = accountName);
+            insert account;
+        } else {
+            account = accounts[0];
+        }
+
+        // List to hold the Opportunities to be upserted
+         // Query existing opportunities for the account
+         Map<String, Opportunity> existingOpportunities = new Map<String, Opportunity>();
+         for (Opportunity opp : [SELECT Id, Name FROM Opportunity WHERE AccountId = :account.Id AND Name IN :oppNames]) {
+             existingOpportunities.put(opp.Name, opp);
+         }
+ 
+         // List to hold the Opportunities to be upserted
+         List<Opportunity> opportunitiesToUpsert = new List<Opportunity>();
+ 
+         // Loop through each Opportunity name
+         for (String oppName : oppNames) {
+             Opportunity opp;
+             if (existingOpportunities.containsKey(oppName)) {
```

Comment:
Very well structured code! Using a Map here helps to streamline the logic and make the method easy to read/understand.